### PR TITLE
381 wire authentication

### PR DESCRIPTION
**Pull request by NhoxxKienn**
## Description
This PR serves to fix the #381 issue of Wire Authentication. 

### TLS Connection
Location: `wire/net/simple`

Issue: currently using TCP without TLS for communication. 

Solution: Implement a simple mutual TLS Dialer and TLS Listener in the `simple` package, the default net wire for `go-perun`.

Bonus: A simple connection test has also been set up with self-signed Certificates in `dialer_internal_test.go`

### Wire identity authentication
Location: `wire/address.go`, `wire/account.go`, and `wire/net/exchange_addr.go`, `wire.proto`

Issue: While each message currently contains a sender and recipient wire address in its Envelope, those are not yet authenticated; there is no signature in the `wire.proto`. This means anyone can send a message coming from SENDER, without needing the private key, thus allowing to spoof any wire identity.

Solution: Add Signing and Verifying for Account and Address. Like Wallet, the Account is expected to hold the private key, and the Address serves as the Public Key. Wire Authentication can be achieved by including a signature signed by the account in the `AuthResponseMsg.` Recipient can check the validity of this signature using the Sender Address (Public Key) to decrypt the signature.

### Wire Authentication Implementation
Location: `backend/sim/wire` and `wire/net/simple`

Description: A simple implementation for the integration of key-pair to be used for wire authentication has been implemented in the testing and default wire. 

For `backend/sim/wire`: This simple signing and verification of signature, independent from the given message.

For `wire/net/simple`: RSA-2048 Random-generated Key-Pair for (Account, Address), Signing of a given message by hashing and encrypting. Signature length: 256 Bits.

An address exchange test has also been added to the package to test the authentication function.

[Visualization](https://miro.com/app/board/uXjVNszT4kY=/?share_link_id=475156674819)

### CODE REVIEW

File: wire/net/simple/address.go [hunk 1]
```diff
@@ -16,54 +16,120 @@ package simple
...
-// returns -1 if a < b, 0 if a == b, 1 if a > b.
-func (a Address) Cmp(b wire.Address) int {
+// Cmp compares the byte representation of two addresses.
+func (a *Address) Cmp(b wire.Address) int {
 	bTyped, ok := b.(*Address)
 	if !ok {
 		panic("wrong type")
 	}
-	return bytes.Compare([]byte(a), []byte(*bTyped))
+	if cmp := bytes.Compare([]byte(a.Name), []byte(bTyped.Name)); cmp != 0 {
```
[the remark this answers is not in the document]
Comment (NhoxxKienn) [reply wire/net/simple/address.go:172]: True, have been reverted to its original implementation.

File: wire/net/simple/address.go [hunk 2]
```diff
@@ -16,54 +16,120 @@ package simple
...
+		Name: host,
+	}
 }
 
 // MarshalBinary marshals the address to binary.
-func (a Address) MarshalBinary() ([]byte, error) {
-	buf := make([]byte, len(a))
-	copy(buf, []byte(a))
-	return buf, nil
+func (a *Address) MarshalBinary() ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (NhoxxKienn) [reply wire/net/simple/address.go:43]: Use custom non-gob encoding

File: wire/net/simple/account.go [hunk 3]
```diff
@@ -15,31 +15,53 @@
...
 func (acc *Account) Address() wire.Address {
 	return acc.addr
 }
 
+// Sign signs the given message with the account's private key.
+func (acc *Account) Sign(msg []byte) ([]byte, error) {
+	if acc.privateKey == nil {
+		return nil, errors.New("private key is nil")
+	}
+	hashed := sha256.Sum256(msg)
```
[the remark this answers is not in the document]
Comment (NhoxxKienn) [reply wire/net/simple/account.go:44]: True, removed hash from CreateAuthRespMsg and VerifyAuthMsg. The address should be in charge of hashing.

File: wire/net/exchange_addr.go [hunk 4]
```diff
@@ -120,3 +137,12 @@ func ExchangeAddrsPassive(ctx context.Context, id wire.Account, conn Conn) (wire
...
 	return addr, err
 }
+
+func verifyAddressSignature(addr wire.Address, sig []byte) error {
+	addressBytes, err := addr.MarshalBinary()
+	if err != nil {
+		return fmt.Errorf("failed to marshal address: %w", err)
+	}
+	hashed := sha256.Sum256(addressBytes)
+	return addr.Verify(hashed[:], sig)
```
[the remark this answers is not in the document]
Comment (NhoxxKienn) [reply wire/net/exchange_addr.go]: Same, move hash to verify for genralization.

File: wire/account.go [hunk 5]
```diff
@@ -31,14 +34,20 @@ func init() {
...
 }
 
 var _ Msg = (*AuthResponseMsg)(nil)
 
 // AuthResponseMsg is the response message in the peer authentication protocol.
 //
 // This will be expanded later to contain signatures.
-type AuthResponseMsg struct{}
+type AuthResponseMsg struct {
+	SignatureSize uint32 // Length of the signature
```
[the remark this answers is not in the document]
Comment (NhoxxKienn) [reply wire/account.go]: True, removed signature_size from AuthResponseMsg

**Review by DragonDev1906 — APPROVED** [review]

**Pull request merged**